freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UsersyncPoller.java

```diff
@@ -0,0 +1,110 @@
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.freeipa.api.v1.freeipa.user.model.SyncOperationStatus;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.entity.UserSyncStatus;
+import com.sequenceiq.freeipa.service.freeipa.user.model.UmsEventGenerationIds;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Service
+public class UsersyncPoller {
+    @VisibleForTesting
+    static final String INTERNAL_ACTOR_CRN = new InternalCrnBuilder(Crn.Service.IAM).getInternalCrnForServiceAsString();
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UsersyncPoller.class);
+
+    @Inject
+    private ThreadBasedUserCrnProvider threadBasedUserCrnProvider;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private UserSyncStatusService userSyncStatusService;
+
+    @Inject
+    private UserService userService;
+
+    @Inject
+    private UmsEventGenerationIdsUtil umsEventGenerationIdsUtil;
+
+    @Value("${freeipa.syncoperation.poller.enabled:true}")
+    private boolean enabled;
+
+    @Scheduled(fixedDelayString = "${freeipa.syncoperation.poller.fixed-delay-millis:60000}",
+            initialDelayString = "${freeipa.syncoperation.poller.initial-delay-millis:300000}")
+    public void pollUms() {
+        try {
+            if (enabled) {
+                syncFreeIpaStacks();
+            }
+        } catch (Exception e) {
+            LOGGER.error("Failed to automatically sync users to FreeIPA stacks", e);
+        }
+    }
+
+    @VisibleForTesting
+    void syncFreeIpaStacks() {
+        threadBasedUserCrnProvider.setUserCrn(INTERNAL_ACTOR_CRN);
+        Optional<String> requestId = Optional.of(UUID.randomUUID().toString());
+        LOGGER.debug("Setting request id = {} for this poll", requestId);
+        MDCBuilder.addRequestId(requestId.get());
+        try {
+            LOGGER.debug("Attempting to sync users to FreeIPA stacks");
+            List<Stack> stackList = stackService.findAllRunning();
+            LOGGER.debug("Found {} active stacks", stackList.size());
+
+            stackList.stream()
+                    .collect(Collectors.groupingBy(Stack::getAccountId))
+                    .entrySet().stream()
+                    .forEach(stringListEntry -> {
+                        UmsEventGenerationIds currentGeneration =
+                                umsEventGenerationIdsUtil.getEventGenerationIds(INTERNAL_ACTOR_CRN, stringListEntry.getKey(), requestId);
+                        stringListEntry.getValue().stream()
+                                .forEach(stack -> {
+                                    if (isStale(stack, currentGeneration)) {
+                                        LOGGER.debug("Environment {} in Account {} is stale.", stack.getEnvironmentCrn(), stack.getAccountId());
+                                        SyncOperationStatus status = userService.synchronizeUsers(stack.getAccountId(), INTERNAL_ACTOR_CRN,
```

Comment:
I want to make sure that I understand the expectations here. This will trigger user sync from every poller right? so if you have 3 instances the expectation is that 2 will fail (as there will be one ongoing sync)? Should you coordinate who runs this code?

Reply:
you are correct.
it would be better to coordinate but I don't think it's possible at the moment.

Reply:
How would Quartz help if we don't have any communication between the FMS instances? If we need to differentiate the instances in some way, then can't we just "disable" the poller in the configuration for 2 of the FMS instances?

I'm not too concerned about this redundancy because we will eventually replace the polling with Cadence. In the meantime, the work performed here is not very heavy compared to the usersync process itself. The first thing user sync does is to accept or reject the operation if there is already a sync running. Also, I do not see any reason for the timing of the three pollers to converge, so the polls will be arbitrarily spread across the polling interval.

If you are really concerned with separating the polling, we can add a table to the db where the pollers can store a lock with and expiration time to guarantee a minimum time between polls. Let me know if you want that. I would like to add that in future work rather than in this commit.

Reply:
So using the DB for synchronizing this is I think pretty easy to do. Chris mentioned to me that you can use k8s to "elect" a master. Basically there is a way to tell k8s to mark one of the containers as master. You can then run the poller from there and only from there. That may be the easiest solution for now.

Reply:
For Kubernetes based leader election you either include a new side container to your POD or you need to invoke Kube endpoints. In autoscaling we implemented leader election based on DB only simply try to lock a certain row and whoever is able to do that is leader. If that one dies the next one who can will be leader. I would avoid leveraging Kubernetes for this, because FMS is not tied to any platform and can run simply on mac in Docker. This is how we use it and i wouldn't tie it to K8S if we don't have to.